File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const COLORS = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white'
] as const

type Color = typeof COLORS[number]

export function decodedResistorValue(colors: Color[]): string {
  let first = String(COLORS.indexOf(colors[0]))
  if (first === "0") {
    first = ""
  }
  
  let second = String(COLORS.indexOf(colors[1]))
  let num = COLORS.indexOf(colors[2])
  if (COLORS.indexOf(colors[1]) == 0) {
    second = ""
    num += 1
  }
  console.log(num)

  let suffix = "0"
  if (num < 3) {
    console.log(suffix)
    console.log(num)
    suffix = suffix.repeat(num) + " "
    console.log(suffix)
  } else if (num < 6) {
    suffix = suffix.repeat(num - 3) + " kilo"
  } else if (num < 9) {
    suffix = suffix.repeat(num - 6) + " mega"
  } else if (num < 12) {
    suffix =suffix.repeat(num - 9) + " giga"
  }

  return `${first}${second}${suffix}ohms`
}

File: solutions/java/resistor-color-trio/1/src/main/java/ResistorColorTrio.java
class ResistorColorTrio {
    static String label(String[] colors) {
        int firstOrdinal = getOrdinal(colors[0]);
        int secondOrdinal = getOrdinal(colors[1]);
        int thirdOrdinal = getOrdinal(colors[2]);
        int countZeroes = secondOrdinal == 0 ? thirdOrdinal + 1 : thirdOrdinal;
        
        return "%s%s%s".formatted(
                getPrefix(firstOrdinal),
                getPrefix(secondOrdinal),
                getSuffix(countZeroes)
        );
    }

    static int getOrdinal(String color) {
        return Colors.valueOf(color.toUpperCase()).ordinal();
    }

    static String getPrefix(int digit) {
        return digit == 0 ? "" : "" + digit;
    }

    static String getSuffix(int ordinal) {
        return switch (ordinal) {
            case 0, 1, 2 -> "0".repeat(ordinal) + " ohms";
            case 3, 4, 5 -> "0".repeat(ordinal - 3) + " kiloohms";
            case 6, 7, 8 -> "0".repeat(ordinal - 6) + " megaohms";
            case 9 -> " gigaohms";
            default -> throw new IllegalStateException("Unexpected value: " + ordinal);
        };
    }
}

enum Colors {
    BLACK,
    BROWN,
    RED,
    ORANGE,
    YELLOW,
    GREEN,
    BLUE,
    VIOLET,
    GREY,
    WHITE
}
